File: apua34/16926.배열돌리기1.js
const fs = require("fs");
const filePath = process.platform === 'linux' ? '/dev/stdin' : 'input.txt';

let input = fs.readFileSync(filePath)
    .toString()
    .trim()
    .split("\n")

const [inputs, ...arr] = input;
let numberArr = [];

arr.map(item => {
    numberArr.push(item.split(' ').map(Number));
});

function rotate(N, M, currentArr) {

    const currentCount = Math.min(N, M) / 2;
    let rotationalArr = currentArr;
  
    for(let rotation = 0; rotation < currentCount; rotation++) {
        const row = N - rotation - 1;
        const col = M - rotation - 1;
        const tmp = rotationalArr[rotation][rotation];
    
        // TOP
        for(let j = rotation; j < col; j++) {
            rotationalArr[rotation][j] = rotationalArr[rotation][j + 1];
        }
        // RIGHT
        for(let i = rotation; i < row; i++) {
            rotationalArr[i][col] = rotationalArr[i + 1][col];
        }
        // BOTTOM
        for(let j = col; j > rotation; j--) {
            rotationalArr[row][j] = rotationalArr[row][j - 1];
        }
        // LEFT
        for(let i = row; i > rotation; i--) {
            rotationalArr[i][rotation] = rotationalArr[i - 1][rotation];
        }

        rotationalArr[rotation + 1][rotation] = tmp;
    }

    return rotationalArr;
}

function solution(inputs, arr) {
    // N : x축, M : y축, R : 회전수
    const [N, M, R] = inputs.split(' ').map(Number);
    let result = arr;
    
    for(let i = 0; i < R; i++) {
        result = rotate(N, M, result);
    }

    return result.map(row => {
        console.log(row.join(' '));
    });
}

solution(inputs, numberArr);

File: Mootata/33회차/16926.배열 돌리기 1.py
from collections import deque

n, m, r = map(int, input().split())
a = [list(input().split()) for _ in range(n)]
q = deque()
answer = [[0 for _ in range(m)] for _ in range(n)]

for i in range(min(n, m) // 2):
    q = deque()
    q.extend(a[i][i : m - i])  # 상
    q.extend([j[m - i - 1] for j in a[i + 1 : n - i - 1]])  # 우
    q.extend(a[n - i - 1][i : m - i][::-1])  # 하
    q.extend([j[i] for j in a[i + 1 : n - i - 1]][::-1])  # 좌

    q.rotate(-r)

    for j in range(i, m - i):  # 위쪽
        answer[i][j] = q.popleft()
    for j in range(i + 1, n - i - 1):  # 오른쪽
        answer[j][m - i - 1] = q.popleft()
    for j in range(m - i - 1, i - 1, -1):  # 아래쪽
        answer[n - i - 1][j] = q.popleft()
    for j in range(n - i - 2, i, -1):  # 왼쪽
        answer[j][i] = q.popleft()

for a in answer:
    print(" ".join(a))

# 1 2 3 4 5 6 7 8
# 1 2 3 4 5 6 7 8
# 1 2 3 4 5 6 7 8
# 1 2 3 4 5 6 7 8
# 1 2 3 4 5 6 7 8
# 1 2 3 4 5 6 7 8
# 1 2 3 4 5 6 7 8
# 1 2 3 4 5 6 7 8
